mempool/src/pool/tests.rs

```diff
@@ -0,0 +1,1787 @@
+    let input = TxInput::new(
+        outpoint_source_id.clone(),
+        0,
+        InputWitness::NoSignature(Some(DUMMY_WITNESS_MSG.to_vec())),
+    );
+    let witness = b"attempted_double_spend".to_vec();
+    let duplicate_input = TxInput::new(
+        outpoint_source_id,
+        0,
+        InputWitness::NoSignature(Some(witness)),
+    );
+    let flags = 0;
+    let locktime = 0;
+    let outputs = tx_spend_input(&mempool, input.clone(), None, flags, locktime)?
+        .outputs()
+        .clone();
+    let inputs = vec![input, duplicate_input];
+    let tx = Transaction::new(flags, inputs, outputs, locktime)?;
+
+    assert!(matches!(
+        mempool.add_transaction(tx),
+        Err(Error::TxValidationError(TxValidationError::DuplicateInputs))
+    ));
+    Ok(())
+}
+
+#[test]
+fn tx_already_in_mempool() -> anyhow::Result<()> {
+    let mut mempool = setup();
+
+    let genesis_tx = mempool
+        .chain_state
+        .confirmed_txs()
+        .values()
+        .next()
+        .expect("genesis tx not found");
+
+    let outpoint_source_id = OutPointSourceId::Transaction(genesis_tx.get_id());
+    let input = TxInput::new(
+        outpoint_source_id,
+        0,
+        InputWitness::NoSignature(Some(DUMMY_WITNESS_MSG.to_vec())),
+    );
+
+    let flags = 0;
+    let locktime = 0;
+    let tx = tx_spend_input(&mempool, input, None, flags, locktime)?;
+
+    mempool.add_transaction(tx.clone())?;
+    assert!(matches!(
+        mempool.add_transaction(tx),
+        Err(Error::TxValidationError(
+            TxValidationError::TransactionAlreadyInMempool
+        ))
+    ));
+    Ok(())
+}
+
+#[test]
+fn outpoint_not_found() -> anyhow::Result<()> {
+    let mut mempool = setup();
+
+    let genesis_tx = mempool
+        .chain_state
+        .confirmed_txs()
+        .values()
+        .next()
+        .expect("genesis tx not found");
+
+    let outpoint_source_id = OutPointSourceId::Transaction(genesis_tx.get_id());
+
+    let good_input = TxInput::new(
+        outpoint_source_id.clone(),
+        0,
+        InputWitness::NoSignature(Some(DUMMY_WITNESS_MSG.to_vec())),
+    );
+    let flags = 0;
+    let locktime = 0;
+    let outputs = tx_spend_input(&mempool, good_input, None, flags, locktime)?.outputs().clone();
+
+    let bad_outpoint_index = 1;
+    let bad_input = TxInput::new(
+        outpoint_source_id,
+        bad_outpoint_index,
+        InputWitness::NoSignature(Some(DUMMY_WITNESS_MSG.to_vec())),
+    );
+
+    let inputs = vec![bad_input];
+    let tx = Transaction::new(flags, inputs, outputs, locktime)?;
+
+    assert!(matches!(
+        mempool.add_transaction(tx),
+        Err(Error::TxValidationError(
+            TxValidationError::OutPointNotFound { .. }
+        ))
+    ));
+
+    Ok(())
+}
+
+#[test]
+fn tx_too_big() -> anyhow::Result<()> {
+    let mut mempool = setup();
+    let tx = TxGenerator::new()
+        .with_num_outputs(400_000)
```

Comment:
I think the issue with this test is that if we ever change the default value, the test can break. I think the right way here is to calculate the size of a single output, then calculate how many outputs are required to break this test based on the limit value.

Reply:
Done.